{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import numpy as np\n",
    "import roman\n",
    "import collections\n",
    "\n",
    "# SET PLOTTING SETTINGS\n",
    "SMALL_SIZE = 16\n",
    "MEDIUM_SIZE = 20\n",
    "BIGGER_SIZE = 24\n",
    "BIGGEST_SIZE = 28\n",
    "\n",
    "plt.rc('font', size=MEDIUM_SIZE)          # controls default text sizes\n",
    "plt.rc('axes', titlesize=MEDIUM_SIZE)     # fontsize of the axes title\n",
    "plt.rc('axes', labelsize=MEDIUM_SIZE)    # fontsize of the x and y labels\n",
    "plt.rc('xtick', labelsize=MEDIUM_SIZE)    # fontsize of the tick labels\n",
    "plt.rc('ytick', labelsize=MEDIUM_SIZE)    # fontsize of the tick labels\n",
    "plt.rc('legend', fontsize=MEDIUM_SIZE)    # legend fontsize\n",
    "plt.rc('figure', titlesize=MEDIUM_SIZE)  # fontsize of the figure title"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_within_window(loc, upstream):\n",
    "    '''Returns true if the sequence loc falls within a window contained in upstream.'''\n",
    "    \n",
    "    start = loc['start']\n",
    "    end = loc['end']\n",
    "\n",
    "    for i,row in upstream.iterrows():\n",
    "        # the end falls into the upstream ORF interval\n",
    "        if end <= row['end'] and end >= row['start']:\n",
    "            return True\n",
    "        # the start falls into the upstream ORF interval\n",
    "        elif start <= row['end'] and start >= row['start']:\n",
    "            return True\n",
    "        # the binding event encompasses the entire upstream ORF interval\n",
    "        elif start <= row['start'] and end >= row['end']:\n",
    "            return True \n",
    "        \n",
    "    return False"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data loading and processing\n",
    "Load our Excel file with the ChIP-exo data, find all genes with SNRs and p-values and save them in dictionaries for each PDM and experiment. Also load Ostrow et al. peak events and make sure they fall in 500 bp windows upstream of ORFs they identified. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "experiments = ['Fkh1_log','Fkh1_stat','Fkh2_log','Fkh2_stat']\n",
    "\n",
    "df_complete = pd.read_pickle('../Data/df_ORF_condition_normalized')\n",
    "\n",
    "targets = {}; MACE = {}; GEM = {}; maxPeak = {}\n",
    "for exp in experiments:\n",
    "    # Load the targets: surpassed threshold in 2 or 3 PDMs\n",
    "    df_targets = pd.read_excel('../Tables/targets_'+exp.replace('_',' ')+'.xlsx', index_col=0)\n",
    "    targets[exp] = df_targets.index.tolist()\n",
    "    \n",
    "    # GEM\n",
    "    df = pd.read_csv('../Data/ORF/GEM_'+exp+'.bed',sep='\\t')\n",
    "    df.columns = ['Chr','start','end','Gene','Gene_std','Score','Motif']\n",
    "    # only genes with a SNR \n",
    "    df = df[df['Score'] > 0] \n",
    "    df = df.set_index('Gene')\n",
    "    df.Chr = df.Chr.map(lambda x: roman.fromRoman(x))\n",
    "    df = df.astype({'start':int,'end':int})\n",
    "    GEM[exp] = df\n",
    "    del(df)\n",
    "    \n",
    "    # MACE \n",
    "    df = pd.read_csv('../Data/ORF/MACE_'+exp+'.bed',sep='\\t')\n",
    "    df.columns = ['Chr','start','end','Gene','Score']\n",
    "    # only genes with a p-value\n",
    "    df = df[df['Score'] > 0]\n",
    "    df = df.set_index('Gene')\n",
    "    df.Chr = df.Chr.map(lambda x: x.replace('chr',''))\n",
    "    df.Chr = df.Chr.map(lambda x: roman.fromRoman(x))\n",
    "    df = df.astype({'start':int,'end':int})\n",
    "    MACE[exp] = df\n",
    "    del(df)\n",
    "    \n",
    "    # maxPeak\n",
    "    df = pd.read_csv('../Data/ORF/condition_normalized/maxPeak_AD_12_'+exp+'.bed',sep='\\t')\n",
    "    df.columns = ['Chr','start','end','Gene','Score']\n",
    "    # get rid of NA's due to zeros\n",
    "    df = df[df['Score'] > 0]\n",
    "    df = df.set_index('Gene')\n",
    "    df.Chr = df.Chr.map(lambda x: x.replace('chr',''))\n",
    "    df.Chr = df.Chr.map(lambda x: roman.fromRoman(x))\n",
    "    df = df.astype({'start':int,'end':int})\n",
    "    maxPeak[exp] = df\n",
    "    del(df)\n",
    "    \n",
    "df_ostrow = pd.read_csv('../Data/Ostrow_et_al_2014/Table_S2.csv')\n",
    "df_ostrow.columns = ['Index','Chr','start','end','class']\n",
    "df_ostrow = df_ostrow[['Chr','start','end','class']]\n",
    "\n",
    "df_ostrow_all_Fkh1 = df_ostrow[df_ostrow['class'].isin(['Fkh1-only','Fkh1and2'])]\n",
    "df_ostrow_all_Fkh2 = df_ostrow[df_ostrow['class'].isin(['Fkh2-only','Fkh1and2'])]\n",
    "\n",
    "# identify those peaks that are within 500 bp windows upstream of ORF starts\n",
    "df_upstream = pd.read_excel('../Data/Ostrow_et_al_2014/Table_S4.xlsx', header=None, \n",
    "                           names=['Chr','start','end','gene','gene_std','Class'])\n",
    "\n",
    "df_ostrow['upstream_orf'] = df_ostrow.apply(lambda x: is_within_window(x, df_upstream[df_upstream.Chr == x['Chr']]),axis=1)\n",
    "df_ostrow = df_ostrow[df_ostrow['upstream_orf']]\n",
    "\n",
    "df_ostrow_Fkh1 = df_ostrow[df_ostrow['class'].isin(['Fkh1-only','Fkh1and2'])]\n",
    "df_ostrow_Fkh2 = df_ostrow[df_ostrow['class'].isin(['Fkh2-only','Fkh1and2'])]\n",
    "df_ostrow = {'Fkh1':df_ostrow_Fkh1,'Fkh2':df_ostrow_Fkh2} "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Count how many of our peak events overlap with Ostrow et al. enriched regions\n",
    "Loop over the experiments, loop over each of the target genes identified in ChIP-exo, search for a peak window in the Ostrow dataset on the same chromosome. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "overlap_counts = {exp:{'Overlap':0,'No overlap':0} for exp in experiments}\n",
    "overlap_counts_all = {exp:{'Overlap':0,'No overlap':0} for exp in experiments}\n",
    "\n",
    "for exp in experiments:\n",
    "    for target in targets[exp]:\n",
    "        std_name = df_complete.loc[target]['Standard name']\n",
    "        \n",
    "        if df_complete.loc[target]['maxPeak_AD_12 '+exp.replace('_',' ')] > 1:\n",
    "            target_row = maxPeak[exp].loc[target]\n",
    "        elif df_complete.loc[target]['GEM '+exp.replace('_',' ')] > 1:\n",
    "            target_row = GEM[exp].loc[target]\n",
    "        \n",
    "        # if there were two peak events for this gene take the most significant one\n",
    "        if not isinstance(target_row,pd.Series):\n",
    "            target_row = target_row[target_row['Score'] == target_row['Score'].max()].iloc[0]\n",
    "            \n",
    "        chromo = target_row.Chr\n",
    "        \n",
    "        # look at overlap upstream of ORFs only\n",
    "        if 'Fkh1' in exp:\n",
    "            overlap = is_within_window(target_row, df_ostrow_Fkh1[df_ostrow_Fkh1.Chr == chromo])\n",
    "        else:\n",
    "            overlap = is_within_window(target_row, df_ostrow_Fkh2[df_ostrow_Fkh2.Chr == chromo])\n",
    "            \n",
    "        if overlap:\n",
    "            overlap_counts[exp]['Overlap'] += 1\n",
    "        else:\n",
    "            overlap_counts[exp]['No overlap'] += 1\n",
    "                \n",
    "        # look at overlap with any enriched region\n",
    "        if 'Fkh1' in exp:\n",
    "            overlap = is_within_window(target_row, df_ostrow_all_Fkh1[df_ostrow_all_Fkh1.Chr == chromo])\n",
    "        else:\n",
    "            overlap = is_within_window(target_row, df_ostrow_all_Fkh2[df_ostrow_all_Fkh2.Chr == chromo])\n",
    "            \n",
    "        if overlap:\n",
    "            overlap_counts_all[exp]['Overlap'] += 1\n",
    "        else:\n",
    "            overlap_counts_all[exp]['No overlap'] += 1\n",
    "    \n",
    "print('Overlap with upstream of ORFs')\n",
    "counts = pd.DataFrame(overlap_counts)\n",
    "counts.at['Total'] = counts.sum()\n",
    "counts = counts.astype(int)\n",
    "counts.at['Fraction'] = counts.apply(lambda x: round(x.Overlap / x.Total, 2))\n",
    "display(counts)\n",
    "\n",
    "print('Overlap with any enriched region')\n",
    "counts_all = pd.DataFrame(overlap_counts_all)\n",
    "counts_all.at['Total'] = counts_all.sum()\n",
    "counts_all = counts_all.astype(int)\n",
    "counts_all.at['Fraction'] = counts_all.apply(lambda x: round(x.Overlap / x.Total,2))\n",
    "display(counts_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Count how many of Ostrow et al.'s enriched regions contain a significant peak event"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate dictionaries of events above the thresholds for each PDM\n",
    "maxPeak_sign = maxPeak.copy()\n",
    "GEM_sign = GEM.copy()\n",
    "MACE_sign = MACE.copy()\n",
    "\n",
    "for exp in experiments:\n",
    "    maxPeak_sign[exp] = maxPeak_sign[exp][maxPeak_sign[exp]['Score'] >= 1]\n",
    "    GEM_sign[exp] = GEM_sign[exp][GEM_sign[exp]['Score'] >= 1]\n",
    "    MACE_sign[exp] = MACE_sign[exp][MACE_sign[exp]['Score'] <= 0.005]\n",
    "\n",
    "\n",
    "results = {exp:0 for exp in experiments}\n",
    "for exp in experiments:\n",
    "    \n",
    "    if 'Fkh1' in exp:\n",
    "        relevant_ostrow_df = df_ostrow_Fkh1\n",
    "    else: \n",
    "        relevant_ostrow_df = df_ostrow_Fkh2\n",
    "        \n",
    "    for i,row in relevant_ostrow_df.iterrows():\n",
    "        chromo = row.Chr\n",
    "\n",
    "        for PDM in [maxPeak_sign[exp],GEM_sign[exp],MACE_sign[exp]]:\n",
    "            contains_peak = is_within_window(row, PDM[PDM.Chr == chromo])\n",
    "\n",
    "            if contains_peak:\n",
    "                results[exp] += 1\n",
    "                break\n",
    "\n",
    "# normalize the counts to fraction of the total number of enriched regions\n",
    "for exp in experiments:\n",
    "    if 'Fkh1' in exp:\n",
    "        relevant_ostrow_df = df_ostrow_Fkh1\n",
    "    else: \n",
    "        relevant_ostrow_df = df_ostrow_Fkh2\n",
    "        \n",
    "    results[exp] = round(results[exp] / len(relevant_ostrow_df),2)\n",
    "\n",
    "# show results\n",
    "pd.DataFrame(results,index=['# enriched ChIP-chip regions containing a ChIP-exo peak'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Target IDs\n",
    "For each experiment (outer key) get locations (Values) on each chromosome (key) for all our targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "chromosomes = list(range(1,17))\n",
    "chromosomes.append(1000) # 1000 is for M, which is the Mitochondrial DNA\n",
    "target_markers_x = {exp:{chromo:[] for chromo in chromosomes} for exp in experiments}\n",
    "target_markers_y = {exp:{chromo:[] for chromo in chromosomes} for exp in experiments}\n",
    "target_markers_ID = {exp:{chromo:[] for chromo in chromosomes} for exp in experiments}\n",
    "\n",
    "for exp in experiments:\n",
    "    for g in targets[exp]:\n",
    "        # get scores for g in exp for each PDM\n",
    "        maxPeak_score = df_complete.loc[g]['maxPeak_AD_12 '+exp.replace('_',' ')]\n",
    "        GEM_score = df_complete.loc[g]['GEM '+exp.replace('_',' ')]\n",
    "        GEM_target = GEM_score >= 1\n",
    "        MACE_score = df_complete.loc[g]['MACE '+exp.replace('_',' ')]\n",
    "        MACE_target = MACE_score <= 0.005\n",
    "        \n",
    "        # Identify y coordinate for target labels \n",
    "        # use the MACE score if not significant in GEM\n",
    "        # or if GEM had a relatively low score compared to MACE\n",
    "        if MACE_target and ((not GEM_target) or (GEM_score <= 4 and MACE_score <= 0.001)  or ((GEM_score <= 2 and MACE_score <= 0.003))):\n",
    "            row = MACE[exp].loc[[g]]\n",
    "            if len(row) > 1:\n",
    "                row = row[row['Score'] == MACE_score]\n",
    "            \n",
    "            target_markers_x[exp][row.Chr[0]].append(row.end[0] + 1)\n",
    "            target_markers_y[exp][row.Chr[0]].append(row.Score[0])\n",
    "            target_markers_ID[exp][row.Chr[0]].append(df_complete.loc[g]['Standard name'])\n",
    "        else: \n",
    "            row = GEM[exp].loc[[g]]\n",
    "            if len(row) > 1:\n",
    "                row = row[row['Score'] == GEM_score]\n",
    "            \n",
    "            # if maxPeak scored higher: replace\n",
    "            if maxPeak_score > GEM_score:\n",
    "                row.Score = maxPeak_score\n",
    "            \n",
    "            target_markers_x[exp][row.Chr[0]].append(row.end[0] + 1)\n",
    "            target_markers_y[exp][row.Chr[0]].append(row.Score[0])\n",
    "            target_markers_ID[exp][row.Chr[0]].append(df_complete.loc[g]['Standard name'])\n",
    "            \n",
    "# fix duplicates by adding their IDs and deleting the second and third etc. occurences\n",
    "for exp in experiments:\n",
    "    for chromo in range(1,17):        \n",
    "        # find duplicate x coordinates\n",
    "        duplicates = [item for item, count in collections.Counter(target_markers_x[exp][chromo]).items() if count > 1]\n",
    "\n",
    "        if len(duplicates) > 0:\n",
    "            # find duplicate indices\n",
    "            dup_indx = [ [i for i,val in enumerate(target_markers_x[exp][chromo]) if val==dup] for dup in duplicates]\n",
    "            for dupl in dup_indx:\n",
    "                # add the gene names\n",
    "                target_markers_ID[exp][chromo][dupl[0]] = ', '.join([target_markers_ID[exp][chromo][i] for i in dupl])\n",
    "                                    \n",
    "            # delete the second, third etc. x, y and ID\n",
    "            # do this in reverse sorted order so that the indices stay the same\n",
    "            to_delete = [dupl[1:] for dupl in dup_indx]\n",
    "            to_delete = sorted([y for x in to_delete for y in x]) # flatten\n",
    "            for idx in reversed(to_delete):\n",
    "                del target_markers_x[exp][chromo][idx]\n",
    "                del target_markers_y[exp][chromo][idx]\n",
    "                del target_markers_ID[exp][chromo][idx]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get all chromosome lengths "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loop over Ostrow, GEM, MACE and maxPeak\n",
    "chromosome_length = {c:0 for c in chromosomes}\n",
    "for TF in ['Fkh1','Fkh2']:\n",
    "    for chromo in chromosomes:\n",
    "        for exp in ['log','stat']:\n",
    "            for (j,df,col) in [(0,df_ostrow[TF],'gray'),(1,GEM[TF+'_'+exp],'seagreen'), \n",
    "                               (2,maxPeak[TF+'_'+exp],'red'), (3,MACE[TF+'_'+exp],'blue')]:\n",
    "\n",
    "                # filter on current chromosome\n",
    "                df_chr = df[df.Chr == chromo] \n",
    "\n",
    "                if len(df_chr) > 0:\n",
    "                    chr_length = int(df_chr['end'].max())\n",
    "\n",
    "                    # keep track of length so that we can set x-range to include all events across all PDMs\n",
    "                    if chr_length > chromosome_length[chromo]:\n",
    "                        chromosome_length[chromo] = chr_length\n",
    "\n",
    "chromosome_length"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate plots\n",
    "Loop over the TFs and chromosomes and plot all PDMs and the Ostrow et al. events."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "plt.figure()    \n",
    "for TF in ['Fkh1','Fkh2']:\n",
    "    for chromo in chromosomes:\n",
    "        for exp in ['log','stat']:\n",
    "            fig, ax1 = plt.subplots()\n",
    "            ax2 = ax1.twinx()\n",
    "\n",
    "            # loop over Ostrow, GEM, MACE and maxPeak\n",
    "            for (PDM,df,col) in [('Ostrow',df_ostrow[TF],'gray'),('maxPeak',maxPeak[TF+'_'+exp],'red'),\n",
    "                               ('GEM',GEM[TF+'_'+exp],'seagreen'),('MACE',MACE[TF+'_'+exp],'blue')]:\n",
    "\n",
    "                # filter on current chromosome\n",
    "                df_chr = df[df.Chr == chromo] \n",
    "                \n",
    "                if len(df_chr) > 0:\n",
    "                    # get x positions \n",
    "                    x = np.array(list(range(chromosome_length[chromo])))\n",
    "                                     \n",
    "                    # plot axvspan's\n",
    "                    # Note that axvspan interprets the ymin and ymax in an interval [0,1], 1 being the top of the plot\n",
    "                    for i,row in df_chr.iterrows():\n",
    "                        if PDM == 'Ostrow':\n",
    "                            # arbitrary value so that the Ostrow peaks reach the top of the plot\n",
    "                            a = 0.25\n",
    "                            ax1.axvspan(xmin=row['start'],xmax=row['end'],ymin=0,ymax=1,color=col,alpha=a)\n",
    "                            \n",
    "                        elif PDM == 'GEM': # GEM\n",
    "                            if row['Score'] >= 1:\n",
    "                                a = 0.8\n",
    "                            else:\n",
    "                                a = 0.5\n",
    "                                \n",
    "                            ymax = np.log2(GEM[TF+'_'+exp].Score.max()) # ~4  \n",
    "                            ylength = ymax + 1 # we start at 2**(-1)\n",
    "                            yloc = np.log2(row['Score']) # where the label is and where the bar ends\n",
    "                            yratio = (yloc + 1)/ylength # in [0,1]\n",
    "                            ax1.axvspan(xmin=row['start'],xmax=row['end'],ymin=0,ymax=yratio,color=col,alpha=a)\n",
    "                        elif PDM == 'maxPeak': # MaxPeak\n",
    "                            if row['Score'] >= 1:\n",
    "                                a = 0.5\n",
    "                            else:\n",
    "                                a = 0.2\n",
    "                            \n",
    "                            ymax = np.log2(GEM[TF+'_'+exp].Score.max()) # ~4  \n",
    "                            ylength = ymax + 1 # we start at 2**(-1)\n",
    "                            yloc = np.log2(row['Score']) # where the label is and where the bar ends\n",
    "                            yratio = (yloc + 1)/ylength # in [0,1]\n",
    "                            ax1.axvspan(xmin=row['start'],xmax=row['end'],ymin=0,ymax=yratio,color=col,alpha=a)\n",
    "                        else:\n",
    "                            if row['Score'] <= 0.005:\n",
    "                                a = 0.8\n",
    "                            else:\n",
    "                                a = 0.5\n",
    "                            \n",
    "                            ymax = -4 # [0.0001,0.01]\n",
    "                            ylength = 2\n",
    "                            yloc = np.log10(row['Score']) # where the label is and where the bar ends\n",
    "                            yratio = -1*(yloc + 2)/ylength # in [0,1]\n",
    "                            ax2.axvspan(xmin=row['start'],xmax=row['end'],ymin=0,ymax=yratio,color=col,alpha=a)\n",
    "                            \n",
    "            # design axes\n",
    "            ax1.set_yscale('log',basey=2)\n",
    "            ax1.set_ylabel(r'$\\log_2$(SNR)')\n",
    "            ax1.yaxis.set_label_position(\"left\")\n",
    "            # set y axis limit based on max GEM score\n",
    "            ax1.set_ylim([0.5,1.1*GEM[TF+'_'+exp].Score.max()])\n",
    "            ax1.set_yticks([2**(int(l)) for l in np.arange(-1,1+np.floor(np.log2(1.1*GEM[TF+'_'+exp].Score.max())),step=1)])\n",
    "            ax1.set_yticklabels([str(int(l)) for l in np.arange(-1,1+np.floor(np.log2(1.1*GEM[TF+'_'+exp].Score.max())),step=1)])\n",
    "            ax1.set_xlabel('base pair location')\n",
    "            ax1.set_xlim([0,chromosome_length[chromo]])\n",
    "            \n",
    "            num_x_labels = int(round(chromosome_length[chromo]/10000))\n",
    "            if num_x_labels < 30:\n",
    "                ax1.set_xticks([i*10000 for i in range(int(round(chromosome_length[chromo]/10000)))])\n",
    "            elif num_x_labels < 45:\n",
    "                ax1.set_xticks([i*25000 for i in range(int(round(chromosome_length[chromo]/25000)))])\n",
    "            else:\n",
    "                ax1.set_xticks([i*50000 for i in range(int(round(chromosome_length[chromo]/50000)))])\n",
    "                \n",
    "            for tick in ax1.get_xticklabels():\n",
    "                tick.set_rotation(45)\n",
    "                \n",
    "            ax2.invert_yaxis()\n",
    "            ax2.set_yscale('log',basey=10)\n",
    "            ax2.set_ylabel(r'$\\log_{10}$(p-value)')\n",
    "            ax2.set_ylim([0.01, 0.0001])\n",
    "            ax2.set_yticks([0.01,0.001, 0.0001])\n",
    "            ax2.set_yticklabels(['-2','-3','-4'])\n",
    "            ax2.set_xlim([0,chromosome_length[chromo]])\n",
    "            ax2.get_xaxis().set_major_formatter(matplotlib.ticker.FuncFormatter(lambda x, p: format(int(x), ','))) # thousands with comma separator\n",
    "\n",
    "            gray_patch = mpatches.Patch(color='gray', label='Ostrow et al. 2014',alpha=0.5)\n",
    "            seagreen_patch = mpatches.Patch(color='seagreen', label='GEM',alpha=0.5)\n",
    "            red_patch = mpatches.Patch(color='red', label=r'$\\it{maxPeak}$',alpha=0.5)\n",
    "            blue_patch = mpatches.Patch(color='blue', label='MACE',alpha=0.5)\n",
    "            ax1.legend(handles=[gray_patch,seagreen_patch,red_patch],loc=2)\n",
    "            ax2.legend(handles=[blue_patch],loc=1)\n",
    "\n",
    "            \n",
    "            # Annotate genes that we consider as targets\n",
    "            # use the SNR value of 1 to see whether we used the GEM/MaxPeak or MACE score for the y value\n",
    "            for g_idx in range(len(target_markers_ID[TF+'_'+exp][chromo])):\n",
    "                if target_markers_y[TF+'_'+exp][chromo][g_idx] >= 1:\n",
    "                    label = target_markers_ID[TF+'_'+exp][chromo][g_idx]\n",
    "                    \n",
    "                    ax1.text(target_markers_x[TF+'_'+exp][chromo][g_idx],\n",
    "                         target_markers_y[TF+'_'+exp][chromo][g_idx],\n",
    "                        target_markers_ID[TF+'_'+exp][chromo][g_idx],fontsize=16)\n",
    "                else: \n",
    "                    label = target_markers_ID[TF+'_'+exp][chromo][g_idx]\n",
    "                    \n",
    "                    ax2.text(target_markers_x[TF+'_'+exp][chromo][g_idx],\n",
    "                         target_markers_y[TF+'_'+exp][chromo][g_idx],\n",
    "                        target_markers_ID[TF+'_'+exp][chromo][g_idx],fontsize=16)\n",
    "                    \n",
    "            # draw cutoff markers on the two y-axes\n",
    "#             m.set_clip_on(False)\n",
    "            ax1.plot(range(chromosome_length[chromo]),[1 for i in range(chromosome_length[chromo])],'--g')\n",
    "            ax2.plot(range(chromosome_length[chromo]),[0.005 for i in range(chromosome_length[chromo])],'--b')\n",
    "#             m.set_clip_on(False)\n",
    "            \n",
    "            plt.title('Chromosome ' + roman.toRoman(chromo))\n",
    "\n",
    "            fig.set_size_inches(25,5)\n",
    "            fig.savefig('../Figures/Ostrow_comparison/'+TF+'_'+exp+'_'+'chr'+str(chromo)+'.png',\n",
    "                        dpi=300, bbox_inches='tight')\n",
    "            plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  },
  "toc": {
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": "block",
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
